File: strategy/custom_monitor/service.js
"use strict";

const MONITOR_DECIMALS = 3;

function round(value, digits = MONITOR_DECIMALS) {
  if (!Number.isFinite(value)) return null;
  const scale = 10 ** digits;
  return Math.round(value * scale) / scale;
}

function toFiniteNumber(value) {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

function normalizeMarket(value, fallback = "A") {
  const market = String(value || fallback).toUpperCase();
  return ["A", "H", "B"].includes(market) ? market : fallback;
}

function normalizeCurrency(value, fallback = "CNY") {
  const currency = String(value || fallback).toUpperCase();
  return ["CNY", "HKD", "USD"].includes(currency) ? currency : fallback;
}

function normalizeSafetyFactor(value, fallback = 1) {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) return fallback;
  if (parsed < 0) return fallback;
  return Math.min(1, parsed);
}

function inferBMarketFromCode(code, fallback = "sh") {
  const normalized = String(code || "").trim();
  if (normalized.startsWith("9")) return "sh";
  if (normalized.startsWith("2")) return "sz";
  return fallback;
}

function hasMonitorTerms(input) {
  return ["stockRatio", "cashDistribution", "cashOptionPrice"]
    .map((key) => Number(input?.[key] || 0))
    .some((value) => Number.isFinite(value) && value !== 0);
}

function toCny(amount, currency, rates) {
  if (!Number.isFinite(amount)) return null;
  const ccy = String(currency || "CNY").toUpperCase();
  if (ccy === "CNY") return amount;
  if (ccy === "HKD") return Number.isFinite(rates?.hkToCny) ? amount * rates.hkToCny : null;
  if (ccy === "USD") return Number.isFinite(rates?.usdToCny) ? amount * rates.usdToCny : null;
  return null;
}

function recalculateMonitor(monitor, rates, nowIso = () => new Date().toISOString()) {
  const targetOriginal = toFiniteNumber(monitor.targetPriceOriginal);
  const targetFallback = toFiniteNumber(monitor.targetPrice);
  const acquirerOriginal = toFiniteNumber(monitor.acquirerPriceOriginal);
  const acquirerFallback = toFiniteNumber(monitor.acquirerPrice);
  const targetPriceCny = toCny(targetOriginal ?? targetFallback, monitor.targetCurrency, rates);
  const acquirerPriceCny = toCny(acquirerOriginal ?? acquirerFallback, monitor.acquirerCurrency, rates);
  const stockRatio = Number(monitor.stockRatio || 0);
  const safetyFactor = normalizeSafetyFactor(monitor.safetyFactor, 1);
  const cashDistributionCny = toCny(Number(monitor.cashDistribution || 0), monitor.cashDistributionCurrency, rates);
  const cashOptionPriceCny = toCny(Number(monitor.cashOptionPrice || 0), monitor.cashOptionCurrency, rates);
  const hasStockLeg = stockRatio !== 0 || Number(monitor.cashDistribution || 0) !== 0;
  const hasCashLeg = Number(monitor.cashOptionPrice || 0) !== 0;
  const stockPayout = hasStockLeg && Number.isFinite(acquirerPriceCny) && Number.isFinite(cashDistributionCny)
    ? acquirerPriceCny * stockRatio * safetyFactor + cashDistributionCny
    : null;
  const stockSpread = Number.isFinite(stockPayout) && Number.isFinite(targetPriceCny)
    ? stockPayout - targetPriceCny
    : null;
  const stockYieldRate = Number.isFinite(stockPayout) && Number.isFinite(targetPriceCny) && targetPriceCny !== 0
    ? (stockSpread / targetPriceCny) * 100
    : null;
  const cashSpread = hasCashLeg && Number.isFinite(cashOptionPriceCny) && Number.isFinite(targetPriceCny)
    ? cashOptionPriceCny - targetPriceCny
    : null;
  const cashYieldRate = hasCashLeg && Number.isFinite(cashOptionPriceCny) && Number.isFinite(targetPriceCny) && targetPriceCny !== 0
    ? (cashSpread / targetPriceCny) * 100
    : null;

  return {
    ...monitor,
    acquirerPrice: round(acquirerPriceCny),
    targetPrice: round(targetPriceCny),
    cashDistributionCny: round(cashDistributionCny),
    cashPayout: hasCashLeg ? round(cashOptionPriceCny) : null,
    stockPayout: round(stockPayout),
    stockSpread: round(stockSpread),
    safetyFactor: round(safetyFactor),
    stockYieldRate: round(stockYieldRate),
    cashSpread: round(cashSpread),
    cashYieldRate: round(cashYieldRate),
    updateTime: nowIso(),
  };
}

function summarizeMonitor(monitor) {
  const stock = Number.isFinite(monitor.stockYieldRate) ? monitor.stockYieldRate : null;
  const cash = Number.isFinite(monitor.cashYieldRate) ? monitor.cashYieldRate : null;
  return {
    id: monitor.id,
    name: monitor.name,
    stockYieldRate: stock,
    cashYieldRate: cash,
    maxYieldRate: stock === null ? cash : cash === null ? stock : Math.max(stock, cash),
  };
}

module.exports = {
  round,
  toFiniteNumber,
  normalizeMarket,
  normalizeCurrency,
  normalizeSafetyFactor,
  inferBMarketFromCode,
  hasMonitorTerms,
  toCny,
  recalculateMonitor,
  summarizeMonitor,
  MONITOR_DECIMALS,
};

File: strategy/custom_monitor/service.py
# AI-SUMMARY: 自定义监控业务计算：组合收益率、对价计算
# 对应 INDEX.md §9 文件摘要索引

"""custom_monitor策略服务。"""

from __future__ import annotations


def to_cny(amount: float | None, currency: str, rates: dict[str, float]) -> float | None:
    """把不同币种统一换算成人民币。"""

    if amount is None:
        return None
    currency_text = str(currency or "CNY").upper()
    if currency_text == "CNY":
        return amount
    if currency_text == "HKD":
        return amount * float(rates.get("hkToCny") or 0) if rates.get("hkToCny") else None
    if currency_text == "USD":
        return amount * float(rates.get("usdToCny") or 0) if rates.get("usdToCny") else None
    return None


def recalculate_monitor(monitor: dict, rates: dict[str, float]) -> dict:
    """按既有公式重算换股腿、现金腿和收益率。"""

    target_price_cny = to_cny(monitor.get("targetPriceOriginal") or monitor.get("targetPrice"), monitor.get("targetCurrency", "CNY"), rates)
    acquirer_price_cny = to_cny(monitor.get("acquirerPriceOriginal") or monitor.get("acquirerPrice"), monitor.get("acquirerCurrency", "CNY"), rates)
    stock_ratio = float(monitor.get("stockRatio") or 0)
    safety_factor = min(1.0, max(0.0, float(monitor.get("safetyFactor") or 1)))
    cash_distribution_cny = to_cny(float(monitor.get("cashDistribution") or 0), monitor.get("cashDistributionCurrency", "CNY"), rates)
    cash_option_price_cny = to_cny(float(monitor.get("cashOptionPrice") or 0), monitor.get("cashOptionCurrency", "CNY"), rates)

    stock_payout = None
    if acquirer_price_cny is not None and cash_distribution_cny is not None and (stock_ratio != 0 or float(monitor.get("cashDistribution") or 0) != 0):
        stock_payout = acquirer_price_cny * stock_ratio * safety_factor + cash_distribution_cny

    stock_spread = (stock_payout - target_price_cny) if stock_payout is not None and target_price_cny not in (None, 0) else None
    stock_yield_rate = ((stock_spread / target_price_cny) * 100) if stock_spread is not None and target_price_cny not in (None, 0) else None
    cash_spread = (cash_option_price_cny - target_price_cny) if cash_option_price_cny is not None and target_price_cny not in (None, 0) else None
    cash_yield_rate = ((cash_spread / target_price_cny) * 100) if cash_spread is not None and target_price_cny not in (None, 0) else None

    return {
        **monitor,
        "acquirerPrice": acquirer_price_cny,
        "targetPrice": target_price_cny,
        "stockSpread": stock_spread,
        "stockYieldRate": stock_yield_rate,
        "cashSpread": cash_spread,
        "cashYieldRate": cash_yield_rate,
    }

